Support for local push in Risco integration

OnFreund opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Currently the Risco integration is reliant on Risco's cloud solution. This PR introduces support for local communication. Both methods have their pros and cons, so the user can pick between the two.

The PyRisco dependency has also been upgraded. This is the only change between the versions:
https://github.com/OnFreund/pyrisco/commit/cfd2f20074f4169f02bd0b52390f5353e3727ba3

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [X] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/23802

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] The code has been formatted using Black (`black --fast homeassistant tests`)
- [X] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [X] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [X] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [X] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [X] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

OnFreund committed:
Local config flow

OnFreund committed:
Local entities

robertboccia commented:
@OnFreund can you tell me is this moving your intergration to local ? if so is this live as im still getting cloud disconnects intermittently? would be great as weve discussed before to have a completely local supported intergration.

pergolafabio commented:
You can choose now between local or cloud with new integration ,... The local is based on same code as the addon, with a few advantages

robertboccia commented:
When will this be officially released?

Robert Boccia
0824971422
________________________________
From: pergolafabio ***@***.***>
Sent: Wednesday, August 3, 2022 9:14:44 PM
To: home-assistant/core ***@***.***>
Cc: Robert Boccia ***@***.***>; Comment ***@***.***>
Subject: Re: [home-assistant/core] Support for local push in Risco integration (PR #75874)


You can choose now between local or cloud with new integration ,... The local is based on same code as the addon, with a few advantages

—
Reply to this email directly, view it on GitHub<https://nam12.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fhome-assistant%2Fcore%2Fpull%2F75874%23issuecomment-1204372074&data=05%7C01%7C%7Cae267e5041bd4d87e19d08da75846d05%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637951508867499502%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=mUQoDO6iavv3uoFKxp07sHTZgFd6eMkDB8jS9B2e1Yc%3D&reserved=0>, or unsubscribe<https://nam12.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAR4IWBH7RV75NMSJXOPS233VXLAKJANCNFSM5456WQTQ&data=05%7C01%7C%7Cae267e5041bd4d87e19d08da75846d05%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637951508867499502%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=J%2BphDZDbstnZIxuK0hT0Wjs7CBzQNr1PjJ5Zo5fzi2k%3D&reserved=0>.
You are receiving this because you commented.Message ID: ***@***.***>

pergolafabio commented:
When it's ready, or just use it now as a custom component

OnFreund commented:
@robertboccia this PR will allow you to choose between cloud and local. It's not live yet, and I don't know when it will be. No one has even started reviewing it yet, so it will probably take a while. I hope it'll be ready for 2022.9, but it doesn't depend on me.

@pergolafabio this is not the same code as the addon, which is in typescript and this is python, but rather the same protocol.

robertboccia commented:
@OnFreund this would be amazing, thank you for all the efforts, this will hopefully bring back the ability to use risco to drive automations i had with lighting etc,

OnFreund commented:
Not sure what's up with the new coverage failure - all I did was rebase against dev...

@robertboccia keep in mind that there are also disadvantages to using local, especially if your board only supports one connection, which means you'll either have to disable cloud, or figure out a proxy configuration (the integration does not provide its own proxy)...

robertboccia commented:
> Not sure what's up with the new coverage failure - all I did was rebase against dev...
> 
> @robertboccia keep in mind that there are also disadvantages to using local, especially if your board only supports one connection, which means you'll either have to disable cloud, or figure out a proxy configuration (the integration does not provide its own proxy)...

I am using a prosysplus so i think the ip module i have is the newer gen that supports cloud and local

MartinHjelmare commented on homeassistant/components/risco/__init__.py:
```diff
@@ -18,17 +29,98 @@
+        raise ConfigEntryNotReady() from error
+    except UnauthorizedError:
+        _LOGGER.exception("Failed to login to Risco cloud")
+        return False
```
A follow up PR should add a reauth flow.

https://developers.home-assistant.io/docs/config_entries_config_flow_handler#reauthentication

MartinHjelmare commented on homeassistant/components/risco/__init__.py:
```diff
@@ -18,17 +29,98 @@
+        _LOGGER.exception("Failed to login to Risco cloud")
+        return False
+
+    async def _error(error):
```
```suggestion
    async def _error(error: Exception) -> None:
```

MartinHjelmare commented on homeassistant/components/risco/__init__.py:
```diff
@@ -18,17 +29,98 @@
+
+    remove_error = risco.add_error_handler(_error)
+
+    async def _default(command, result, *params):
```
Please add type annotations if not too hard.

MartinHjelmare commented on homeassistant/components/risco/__init__.py:
```diff
@@ -18,17 +29,98 @@
+
+    remove_partition = risco.add_partition_handler(_partition)
+
+    listenrs = [remove_error, remove_default, remove_zone, remove_partition]
+
+    listenrs.append(entry.add_update_listener(_update_listener))
+
+    hass.data.setdefault(DOMAIN, {})
+    hass.data[DOMAIN][entry.entry_id] = {
+        UNDO_LISTENERS: listenrs,
```
```suggestion
    listeners = [remove_error, remove_default, remove_zone, remove_partition]

    listeners.append(entry.add_update_listener(_update_listener))

    hass.data.setdefault(DOMAIN, {})
    hass.data[DOMAIN][entry.entry_id] = {
        UNDO_LISTENERS: listeners,
```

MartinHjelmare commented on homeassistant/components/risco/__init__.py:
```diff
@@ -18,17 +29,98 @@
+    async def _error(error):
+        _LOGGER.error("Error in Risco library: %s", error)
+
+    remove_error = risco.add_error_handler(_error)
```
We can use `config_entry.async_on_unload` to call the unsubscribe callbacks.

MartinHjelmare commented on homeassistant/components/risco/alarm_control_panel.py:
```diff
@@ -53,26 +56,42 @@ async def async_setup_entry(
+        async_add_entities(local_entities, False)
+    else:
+        coordinator = hass.data[DOMAIN][config_entry.entry_id][DATA_COORDINATOR]
+        cloud_entities = [
+            RiscoCloudAlarm(
+                coordinator, partition_id, config_entry.data[CONF_PIN], options
+            )
+            for partition_id in coordinator.data.partitions
+        ]
+        async_add_entities(cloud_entities, False)
```
```suggestion
        async_add_entities(
            RiscoCloudAlarm(
                coordinator, partition_id, config_entry.data[CONF_PIN], options
            )
            for partition_id in coordinator.data.partitions
        )
```

MartinHjelmare commented on homeassistant/components/risco/alarm_control_panel.py:
```diff
@@ -53,26 +56,42 @@ async def async_setup_entry(
+    if is_local(config_entry):
+        partition_updates = hass.data[DOMAIN][config_entry.entry_id][PARTITION_UPDATES]
+        system = hass.data[DOMAIN][config_entry.entry_id][SYSTEM]
+        local_entities = [
+            RiscoLocalAlarm(
+                system.id,
+                partition_id,
+                partition,
+                partition_updates,
+                config_entry.data[CONF_PIN],
+                options,
+            )
+            for partition_id, partition in system.partitions.items()
+        ]
+        async_add_entities(local_entities, False)
```
```suggestion
        async_add_entities(
            RiscoLocalAlarm(
                system.id,
                partition_id,
                partition,
                partition_updates,
                config_entry.data[CONF_PIN],
                options,
            )
            for partition_id, partition in system.partitions.items()
        )
```

MartinHjelmare commented on homeassistant/components/risco/__init__.py:
```diff
@@ -18,17 +29,98 @@
+        ZONE_UPDATES: zone_updates,
+        PARTITION_UPDATES: partition_updates,
+        SYSTEM: risco,
+    }
```
I recommend using a dataclass to store multiple instances in hass.data under a single key. That gives type annotations for free.

OnFreund replied:
What do you mean for free? Wouldn't the type definitions simply move to the class?

MartinHjelmare replied:
The attributes will be typed once we know that we've got this dataclass from hass.data. Instead of needing to type all the items all the time, after looking up hass.data, we just need to type the dataclass after looking that up in hass.data.

MartinHjelmare commented on homeassistant/components/risco/alarm_control_panel.py:
```diff
@@ -82,8 +101,10 @@ def __init__(self, coordinator, partition_id, code, options):
+    @property
+    def unique_id(self) -> str:
+        """Return a unique id for this partition."""
+        raise NotImplementedError
```
Nowadays we prefer using our short hand attributes, eg `_attr_unique_id`.

https://developers.home-assistant.io/docs/core/entity#entity-class-or-instance-attributes

OnFreund replied:
Cool. What about the abstract base class? Do I simply remove this property?

MartinHjelmare replied:
Yeah, I'd do that. It's only ourselves that will be using the base class so it's not that fragile.

OnFreund replied:
This causes a new problem though:
```
error: Argument 1 to <set> has incompatible type "Tuple[str, Optional[str]]"; expected "Tuple[str, str]"  [arg-type]
```

This is for the call to `self.unique_id` from `device_info()`

MartinHjelmare replied:
Maybe pass the device unique_id as a parameter to this class and set `_attr_device_info` in init method?

OnFreund replied:
That will require every class to do that, and we might have new ones coming up when we split `bypassed` and `alarmed`.
Maybe I can initialize to an empty string in the base class, and use that instead of `self.unique_id`?

MartinHjelmare replied:
Setting an empty string doesn't sound correct.

If the different children represent different devices with different unique ids it seems valid that they either provide the unique id to the parent which sets the device info, or that the children themselves set the device info.

OnFreund replied:
Yeah, I see your point about empty string not sounding right, but it's a bummer that we're adding additional coupling points just to make typing hints work in a situation where there are no real typing concerns.

OnFreund replied:
it's also a cycle because the unique id depends on `self.risco` which is set by one of the base class constructors. Argghh

MartinHjelmare replied:
Let's see how it looks after the next push. It's easier to make suggestions then.

OnFreund replied:
Fair enough. Coming up. I had to add `or ''` just so I can commit

MartinHjelmare commented on homeassistant/components/risco/binary_sensor.py:
```diff
@@ -82,15 +88,71 @@ def device_class(self):
+    def __init__(self, coordinator, zone_id, zone):
+        """Init the zone."""
+        RiscoEntity.__init__(self, coordinator)
+        RiscoBinarySensor.__init__(self, zone_id, zone)
```
It's better to pass on not used parameters to the parent via `super()` than calling different parents specifically.

OnFreund replied:
Not sure I understand - is there an example?

MartinHjelmare replied:
I'd do it like this:

```py
class RiscoEntity(CoordinatorEntity[RiscoDataUpdateCoordinator]):
    """Risco entity base class."""

    ...

...


class RiscoBinarySensor(BinarySensorEntity):
    """Representation of a Risco zone as a binary sensor."""

    def __init__(self, *, zone_id: int, zone: Zone, **kwargs: Any) -> None:
        """Init the zone."""
        super().__init__(**kwargs)
        self._zone_id = zone_id
        self._zone = zone

    ...


class RiscoCloudBinarySensor(RiscoBinarySensor, RiscoEntity):
    """Representation of a Risco cloud zone as a binary sensor."""

    def __init__(
        self, coordinator: RiscoDataUpdateCoordinator, zone_id: int, zone: Zone
    ) -> None:
        """Init the zone."""
        super().__init__(zone_id=zone_id, zone=zone, coordinator=coordinator)

    ...


class RiscoLocalBinarySensor(RiscoBinarySensor):
    """Representation of a Risco local zone as a binary sensor."""

    def __init__(
        self,
        system_id: str,
        zone_id: int,
        zone: Zone,
        zone_updates: dict[int, Callable[[], Any]],
    ) -> None:
        """Init the zone."""
        super().__init__(zone_id=zone_id, zone=zone)
        self._system_id = system_id
        self._zone_updates = zone_updates

    ...
```

OnFreund replied:
thank you!

MartinHjelmare commented on homeassistant/components/risco/binary_sensor.py:
```diff
@@ -82,15 +88,71 @@ def device_class(self):
+        return {
+            **super().extra_state_attributes,
+            "alarmed": self._zone.alarmed,
+            "groups": self._zone.groups,
```
What do these state attributes represent? We try to avoid adding extra state attributes as far as possible.

OnFreund replied:
`alarmed` is whether this zone is currently causing an alarm, `groups` are the arming groups the zone belongs to. The former can arguably be a separate binary sensor, the latter is fairly static (requires reprogramming the alarm, which typically only installers have access to).
We already have `bypassed`, which today would have been a switch, but requires some restructuring (and is a breaking change), so I suggest I split off `bypassed` and `alarmed` to separate entities in a future PR.

MartinHjelmare replied:
How will the users use the `groups` info?

We shouldn't add `alarmed` now if we plan to make it a binary sensor. Either make it a binary sensor now or later.

OnFreund replied:
* I imagine `groups` will most likely be consumed in the UI as a guide when building automations, e.g. if I arm group C, will I need to bypass this zone first?
BTW the fact that we're exposing groups is really cool - the Risco app doesn't show them anywhere, which means people can't even understand how their home is configured.
* I'll probably drop `alarmed` for now and add it in a future PR where I also break `bypassed` out to a switch. Need to first figure out the best way to have multiple entities subscribe to the same zone update.

MartinHjelmare replied:
If the user wants to bypass a zone in the automation based on which group is armed, shouldn't the info relationship be the other way around? Ie we expose the zone entity ids that belong to an alarm group on the alarm entity that controls the alarm group. Then it's easy to check the state attributes of the alarm that is targeted in the automation and it should work to bypass all those zone entities programmatically in the automation.

OnFreund replied:
There are no entities for groups. Groups are just sets of zones, but partitions are the units that are represented by individual alarm control panels (Risco terminology is not ideal).
To add to the complexity, groups span across partitions so are not part of a hierarchy (zones can also belong to multiple partitions).

MartinHjelmare replied:
Ok.

MartinHjelmare commented on homeassistant/components/risco/binary_sensor.py:
```diff
@@ -82,15 +88,71 @@ def device_class(self):
+    @property
+    def should_poll(self):
+        """Return True if entity has to be polled for state."""
+        return False
```
We can set `_attr_should_poll` as a class attribute above the init method.

MartinHjelmare commented on homeassistant/components/risco/config_flow.py:
```diff
@@ -62,6 +73,21 @@ async def validate_input(hass: core.HomeAssistant, data):
     return {"title": risco.site_name}
 
 
+async def validate_local_input(hass: core.HomeAssistant, data):
```
```suggestion
async def validate_local_input(
    hass: core.HomeAssistant, data: Mapping[str, str]
) -> dict[str, str]:
```

MartinHjelmare commented on homeassistant/components/risco/config_flow.py:
```diff
@@ -95,7 +128,32 @@ async def async_step_user(self, user_input=None):
+        """Configure a local based alarm."""
+        errors = {}
+        if user_input is not None:
+            await self.async_set_unique_id(user_input[CONF_HOST])
```
Host is not a valid unique id.

https://developers.home-assistant.io/docs/config_entries_config_flow_handler#unique-id-requirements

OnFreund replied:
ok. I'll change it to the site id, but that requires connecting and querying for it, which in most cases would make this meaningless as Risco usually does not accept another connect (and in many cases it might just get stuck).

This also leads me to a different question I don't know where to ask:
Because we're connecting from the config flow, and then connecting again when setting up the platforms, Risco sometimes gets stuck. Adding `await asyncio.sleep(1)` before creating the config entry solves this, but I don't like it, and I definitely don't want tests to have to wait for it. Any suggestions how to resolve?

MartinHjelmare replied:
Store the sleep time in a constant to make it easy to patch to 0 in tests.

MartinHjelmare commented on tests/components/risco/test_config_flow.py:
```diff
@@ -82,33 +95,133 @@ async def test_form(hass):
+        await hass.async_block_till_done()
+
+    expected_data = {**TEST_LOCAL_DATA, **{"type": "local"}}
+    assert result3["type"] == "create_entry"
```
Please use `FlowResultType` instead.

MartinHjelmare reviewed (changes requested):
Looks good! I recommend adding type annotations to all new and changed signatures. That will make it easier to review the code.

OnFreund committed:
Apply suggestions from code review

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

OnFreund committed:
Address code review comments

OnFreund commented:
Documenting the required follow up PRs for future reference:
* Split `bypassed` into a switch (both cloud and local), and add the `alarmed` binary sensor (local only) [1]
* Add a reauth flow
* Split the options flow according to cloud vs local (adding a couple of local-specific options, and making sure that scan interval is only displayed in cloud)

[1] on second thought, the most common use case for `alarmed` would be for an automation to find the zone that's causing the alarm. Is a separate sensor the best way to support this? How would we link back to the "main" binary sensor? And how would we wait for a trigger on any zone?

OnFreund committed:
More type hints

MartinHjelmare commented on homeassistant/components/risco/__init__.py:
```diff
@@ -18,17 +31,94 @@
+
+    local_data = LocalData(risco)
+
+    async def _zone(zone_id: int, zone: Zone):
```
Don't forget the return value typing.
```suggestion
    async def _zone(zone_id: int, zone: Zone) -> None:
```

MartinHjelmare commented on homeassistant/components/risco/alarm_control_panel.py:
```diff
@@ -53,26 +58,48 @@ async def async_setup_entry(
+            for partition_id, partition in local_data.system.partitions.items()
+        )
+    else:
+        coordinator = hass.data[DOMAIN][config_entry.entry_id][DATA_COORDINATOR]
```
```suggestion
        coordinator: RiscoDataUpdateCoordinator = hass.data[DOMAIN][config_entry.entry_id][
            DATA_COORDINATOR
        ]
```

MartinHjelmare commented on homeassistant/components/risco/alarm_control_panel.py:
```diff
@@ -23,6 +27,7 @@
 from homeassistant.helpers.entity import DeviceInfo
 from homeassistant.helpers.entity_platform import AddEntitiesCallback
 
+from . import is_local
```
```suggestion
from . import LocalData, RiscoDataUpdateCoordinator, is_local
```

MartinHjelmare commented on homeassistant/components/risco/alarm_control_panel.py:
```diff
@@ -53,26 +58,48 @@ async def async_setup_entry(
-
-    async_add_entities(entities, False)
+    if is_local(config_entry):
+        local_data = hass.data[DOMAIN][config_entry.entry_id]
```
```suggestion
        local_data: LocalData = hass.data[DOMAIN][config_entry.entry_id]
```

MartinHjelmare commented on homeassistant/components/risco/alarm_control_panel.py:
```diff
@@ -82,28 +109,15 @@ def __init__(self, coordinator, partition_id, code, options):
-        self._partition = self.coordinator.data.partitions[self._partition_id]
-
     @property
     def device_info(self) -> DeviceInfo:
```
We can set `_attr_device_info`.

MartinHjelmare replied:
Consider setting the device info in the child classes instead as it seems these entities represent the main device feature and have different names and ids for the devices.

MartinHjelmare commented on homeassistant/components/risco/alarm_control_panel.py:
```diff
@@ -165,7 +179,62 @@ async def _arm(self, mode, code):
+            options=options,
+        )
+        self._attr_unique_id = f"{self._risco.site_uuid}_{partition_id}"
+        self._attr_name = f"Risco {self._risco.site_name} Partition {partition_id}"
```
If this represents the main device feature of this device we should set the entity name to `None` while setting `_attr_has_entity_name` to `True`.

https://developers.home-assistant.io/docs/core/entity#entity-naming

MartinHjelmare commented on homeassistant/components/risco/alarm_control_panel.py:
```diff
@@ -165,7 +179,62 @@ async def _arm(self, mode, code):
+    """Representation of a Risco partition."""
+
+    def __init__(
+        self, coordinator, partition_id, code: str, options: dict[str, Any]
```
Please add type annotations for the whole signature.

MartinHjelmare commented on tests/components/risco/test_binary_sensor.py:
```diff
@@ -7,23 +7,30 @@
+    setup_risco_cloud,
+    setup_risco_local,
+    two_zone_cloud,
+    two_zone_local,
```
Don't import pytest fixtures. Either define them in the same module if they are only used for the tests of that module, or in a `conftest.py` module in the package if they are used by more than one module. The fixtures are then available by default to the tests below in the tree.

MartinHjelmare commented on tests/components/risco/util.py:
```diff
@@ -43,15 +58,36 @@ async def setup_risco(hass, events=[], options={}):
     return config_entry
 
 
+async def setup_risco_local(hass, options={}):
```
Don't set a default arguments to a mutable object. That's error prone.

This function can be made a pytest fixture instead that depends on another fixture for the options. We can override fixtures per test via `pytest.mark.parametrize`. Use that to set custom options.

OnFreund replied:
Thanks.
How would you handle situations where we need to make assertions before setup? (e.g. https://github.com/home-assistant/core/pull/75874/files#diff-2a7abdb0037958212636f0163d689d80071e1a19f4946380cf12b81b07231534R65)

Also is there an example of a fixture dependent on another fixture? Not even sure what to search for.

MartinHjelmare replied:
A fixture depends on another fixture if we pass the name of the fixture to depend on as an argument to the fixture that needs the other fixture. This is just standard pytest behavior. Please read more about that in the pytest docs.

Some cases may need the fixture to return or yield a function that can be used in the test as needed.

We often don't assert the lack of states before setting up an integration, though. Unless I'm doing some modifications to the entity or device state before setting up the integration, I'd skip asserting that there are no states at that point. An example where it could be interesting to assert the entity state before setting up the integration is if I'm enabling an entity that is normally disabled before setting up the integration.

I recommend making a separate fixture that returns a mock config entry. Then it's easy to set up the integration using that config entry, inside the test for the cases where we want more control.

OnFreund replied:
Sorry, should have made my question clearer. I know how to pass fixtures by name, and I know how to pass them dynamically via `pytest.lazy_fixture` (for which I couldn't find any reference in the HA codebase, so not sure whether it's frowned upon, or what are the best practices). What I don't know is how to *dynamically* pass a fixture to another one.

I removed the no-state assertions, but there are still other places where code is executed / functions are patched before calling `setup_*`.

In any case, I removed the mutable default arguments through `*args` and `**kwargs` for now, hope that it's satisfactory.

MartinHjelmare replied:
Please explain the problem with a specific example.

OnFreund replied:
For example, [here](https://github.com/home-assistant/core/pull/75874/files#diff-79db24431d93f5c3a4de758b472607e4d6709b8b4b4037767cfeb7443f7a5b37R170) we have:
```python
hass.config.set_time_zone("UTC")
with patch(
        "homeassistant.components.risco.Store.async_save",
    ) as save_mock:
    await setup_risco_cloud(hass, *TEST_EVENTS)
...
```
We have 3 things as a dependency to `setup_risco_cloud`:
1. Setting the time zone
2. Patching `async_save`
3. The test events

MartinHjelmare replied:
The order of a fixture arguments in the test signature will determine the order of fixture setup, unless the fixtures themselves have further dependencies that control the order.

Does that answer your question?

OnFreund replied:
It answers 1 & 2. As for 3, I guess that the latest call to patch is the one that determines the behavior? So if I patch from within `setup_risco_cloud` to return an empty list as events, and then have another fixture in a later argument that patches to return the test events it could work.
I'll try it later

MartinHjelmare replied:
I'd extract the test events to a separate fixture. Then those can be overridden as needed per test.

OnFreund replied:
Patching the events with an empty list is needed everywhere `setup_risco_cloud` is used (with the exception of cases where we're patching with actual data), so it means adding that separate fixture to almost every test. Would the patch over patch approach not work?

MartinHjelmare replied:
I'd make one fixture for `setup_risco_cloud` that patches and sets up the config entry and another fixture that returns the return value for the patch of the events. The cloud setup fixture can depend on the events fixture. The events fixture can return an empty list by default.

OnFreund replied:
OK, so that's exactly the question I previously had. Let's say we have a fixture called `events`, and we use the following signatures:
```python
@fixture
def events(*events):

@fixture
async def setup_risco_cloud(hass, events):
```

If I want to have a test with non-default events:
```python
async def test_cloud_setup(hass, two_zone_cloud, setup_risco_cloud):
```

How do I customize the events, given that the dependency on `events` is not in this signature, but rather in `setup_risco_cloud`?

MartinHjelmare replied:
Use `pytest.mark.parametrize`. The fixture doesn't need to be present in the signature of the current test to override it by parametrization.

MartinHjelmare replied:
Here's one example where we don't even define the `platforms` fixture function. We just override it per test.

https://github.com/home-assistant/core/blob/58b9785485af4b49097707edb7fbcc00c72a3df0/tests/components/evil_genius_labs/conftest.py#L39-L55

https://github.com/home-assistant/core/blob/58b9785485af4b49097707edb7fbcc00c72a3df0/tests/components/evil_genius_labs/test_light.py#L15-L16

OnFreund replied:
Perfect - just what I was looking for

MartinHjelmare commented:
Please go over the code and complete the type annotations for all the changes. There are many missing.

OnFreund committed:
Apply suggestions from code review

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

MartinHjelmare commented:
> How would we link back to the "main" binary sensor?

Isn't it enough to link the binary sensors via the device? They should belong to the same device, right?

> And how would we wait for a trigger on any zone?

Finding entities for different standard categories in Home Assistant, eg integration, area, or device, can be done with our template helpers and then extra filtering as required.

https://www.home-assistant.io/docs/configuration/templating/#integrations
https://www.home-assistant.io/docs/configuration/templating/#areas
https://www.home-assistant.io/docs/configuration/templating/#devices

OnFreund committed:
More annotations

OnFreund committed:
Even more annonations

OnFreund committed:
New entity naming

OnFreund committed:
Move fixtures to conftest

OnFreund commented:
> Isn't it enough to link the binary sensors via the device? They should belong to the same device, right?

Yes, but it would require more templating vs triggering on an attribute and then having access to the entity.

> Finding entities for different standard categories in Home Assistant, eg integration, area, or device, can be done with our template helpers and then extra filtering as required.

Cool, but there are still gaps (or maybe stuff I'm not aware of):
1. None of these seem to be filters. So after getting the list of zones in an area, you still have to filter by integration, or vice versa.
2. How do you trigger based on this? You'll have to combine them into one value, and then once its triggered, do some more templating to extract the actual zone, which by then could change state.

MartinHjelmare commented:
Those are general automation questions/problems. We don't change our entity design for a single integration to try to make a particular automation easier. We follow our guidelines. If there's a general problem we solve it in a general way for all integrations.

OnFreund committed:
Improve state tests for local

OnFreund committed:
Remove mutable default arguments

OnFreund committed:
Remove assertions for lack of state

OnFreund committed:
Add missing file

OnFreund committed:
Switch setup to fixtures

OnFreund committed:
Use error fixtures in test_config_flow

MartinHjelmare commented on homeassistant/components/risco/alarm_control_panel.py:
```diff
@@ -165,7 +174,74 @@ async def _arm(self, mode, code):
         else:
             await self._call_alarm_method(risco_state)
 
+    async def _call_alarm_method(self, method, *args):
```
Type annotation?

MartinHjelmare replied:
```suggestion
    async def _call_alarm_method(self, method: str, *args: Any) -> None:
```

MartinHjelmare commented on homeassistant/components/risco/alarm_control_panel.py:
```diff
@@ -165,7 +174,74 @@ async def _arm(self, mode, code):
+            manufacturer="Risco",
+        )
+
+    def _get_data_from_coordinator(self):
```
```suggestion
    def _get_data_from_coordinator(self) -> None:
```

MartinHjelmare commented on homeassistant/components/risco/alarm_control_panel.py:
```diff
@@ -165,7 +174,74 @@ async def _arm(self, mode, code):
+        """Subscribe to updates."""
+        self._partition_updates[self._partition_id] = self.async_write_ha_state
+
+    async def _call_alarm_method(self, method, *args):
```
```suggestion
    async def _call_alarm_method(self, method: str, *args: Any) -> None:
```

MartinHjelmare commented on homeassistant/components/risco/binary_sensor.py:
```diff
@@ -28,69 +36,119 @@ async def async_setup_entry(
-class RiscoBinarySensor(BinarySensorEntity, RiscoEntity):
+    if is_local(config_entry):
+        local_data: LocalData = hass.data[DOMAIN][config_entry.entry_id]
+        local_entities = [
+            RiscoLocalBinarySensor(
+                local_data.system.id, zone_id, zone, local_data.zone_updates
+            )
+            for zone_id, zone in local_data.system.zones.items()
+        ]
+        async_add_entities(local_entities, False)
```
```suggestion
        async_add_entities(
            RiscoLocalBinarySensor(
                local_data.system.id, zone_id, zone, local_data.zone_updates
            )
            for zone_id, zone in local_data.system.zones.items()
        )
```

MartinHjelmare commented on homeassistant/components/risco/binary_sensor.py:
```diff
@@ -28,69 +36,119 @@ async def async_setup_entry(
+        coordinator: RiscoDataUpdateCoordinator = hass.data[DOMAIN][
+            config_entry.entry_id
+        ][DATA_COORDINATOR]
+        cloud_entities = [
+            RiscoCloudBinarySensor(coordinator, zone_id, zone)
+            for zone_id, zone in coordinator.data.zones.items()
+        ]
+        async_add_entities(cloud_entities, False)
```
```suggestion
        async_add_entities(
            RiscoCloudBinarySensor(coordinator, zone_id, zone)
            for zone_id, zone in coordinator.data.zones.items()
        )
```

MartinHjelmare commented on homeassistant/components/risco/binary_sensor.py:
```diff
@@ -28,69 +36,119 @@ async def async_setup_entry(
+        """Unbypass this zone."""
+        await self._bypass(False)
+
+    async def _bypass(self, bypass):
```
```suggestion
    async def _bypass(self, bypass: bool) -> None:
```

MartinHjelmare commented on homeassistant/components/risco/binary_sensor.py:
```diff
@@ -28,69 +36,119 @@ async def async_setup_entry(
+            name=self._zone.name,
+        )
+
+    def _get_data_from_coordinator(self):
```
```suggestion
    def _get_data_from_coordinator(self) -> None:
```

MartinHjelmare commented on homeassistant/components/risco/binary_sensor.py:
```diff
@@ -28,69 +36,119 @@ async def async_setup_entry(
+    def _get_data_from_coordinator(self):
+        self._zone = self.coordinator.data.zones[self._zone_id]
 
     async def _bypass(self, bypass):
```
```suggestion
    async def _bypass(self, bypass: bool) -> None:
```

MartinHjelmare commented on homeassistant/components/risco/binary_sensor.py:
```diff
@@ -28,69 +36,119 @@ async def async_setup_entry(
+            "groups": self._zone.groups,
+        }
+
+    async def _bypass(self, bypass):
```
```suggestion
    async def _bypass(self, bypass: bool) -> None:
```

MartinHjelmare commented on tests/components/risco/conftest.py:
```diff
@@ -0,0 +1,148 @@
+@fixture
+def options():
+    """Fixture for default (empty) options."""
+    yield {}
```
We only need to yield if we're inside a context manager or if we want to tear down something after yielding.
```suggestion
    return {}
```

MartinHjelmare commented on tests/components/risco/conftest.py:
```diff
@@ -0,0 +1,148 @@
+@fixture
+def events():
+    """Fixture for default (empty) events."""
+    yield []
```
```suggestion
    return []
```

MartinHjelmare commented on tests/components/risco/conftest.py:
```diff
@@ -0,0 +1,148 @@
+        domain=DOMAIN, data=TEST_CLOUD_CONFIG, options=options
+    )
+    config_entry.add_to_hass(hass)
+    yield config_entry
```
```suggestion
    return config_entry
```

MartinHjelmare commented on tests/components/risco/conftest.py:
```diff
@@ -0,0 +1,148 @@
+        await hass.config_entries.async_setup(cloud_config_entry.entry_id)
+        await hass.async_block_till_done()
+
+    return cloud_config_entry
```
We should probably yield the config entry within the patch context manager to keep it applied.

```suggestion
        yield cloud_config_entry
```

MartinHjelmare commented on tests/components/risco/conftest.py:
```diff
@@ -0,0 +1,148 @@
+        domain=DOMAIN, data=TEST_LOCAL_CONFIG, options=options
+    )
+    config_entry.add_to_hass(hass)
+    yield config_entry
```
```suggestion
    return config_entry
```

MartinHjelmare commented on tests/components/risco/conftest.py:
```diff
@@ -0,0 +1,148 @@
+        await hass.config_entries.async_setup(local_config_entry.entry_id)
+        await hass.async_block_till_done()
+
+    yield local_config_entry
```
```suggestion
        yield local_config_entry
```

OnFreund committed:
Apply suggestions from code review

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

MartinHjelmare reviewed (approved):
Thanks!

OnFreund commented:
Thanks so much @MartinHjelmare. Not only is the code way better than what it was, your review also helped get me up to speed on current HA best practices, and I even picked up a few Python tricks along the way!
